Project wide groups implementation

xDGameStudios opened the pull request:
This PR implements global groups as requested by #5744 taking into consideration #18861 and #20416.

The system ended up to look like the mockups in #20416

In the project settings under "Groups" it is possible to:
- create groups (name and description)
- delete groups (this will also remove references to that group in every node in every scene)
- rename groups (this will change the name and fix group references on every node in every scene)
- edit groups description
- groups are auto ordered alphabetically
- You have a confirm notification upon deleting a group

<img width="892" alt="captura de ecra 2018-12-14 as 11 28 28" src="https://user-images.githubusercontent.com/10358443/50001020-6d126e80-ff93-11e8-860a-af64c3af84b5.png">

<img width="639" alt="captura de ecra 2018-12-15 as 19 18 53" src="https://user-images.githubusercontent.com/10358443/50046644-67ee1600-009e-11e9-8b66-50cdcb0c473c.png">

In node groups editor dock it is possible to:
- add node to existent group (checkbox)
- remove node from group (checkbox)
- call the group manager (opens the project settings "Groups" tab)

<img width="373" alt="captura de ecra 2018-12-10 as 15 15 15" src="https://user-images.githubusercontent.com/10358443/49743411-8c0eb900-fc92-11e8-9ce6-e0cfd497fc00.png">

In code editor you get:
- autocompletion for "add_to_group", "remove_from_group" and "is_in_group"

In the plugin editor it is now possible to:
- add groups (create_group(name, description));
- remove groups (delete_group(name));

<img width="477" alt="captura de ecra 2018-12-15 as 19 18 06" src="https://user-images.githubusercontent.com/10358443/50046642-573da000-009e-11e9-87f4-29c60caf9521.png">

NEW addition: groups are now saved as an array of dictionaries to allow for adding new groups related features in the future.

xDGameStudios committed:
Global groups implementation

willnationsdev commented on editor/editor_group_settings.cpp:
```diff
@@ -0,0 +1,745 @@
+
+            Ref<PackedScene> data = ResourceLoader::load(path);
+
+            PackedScene::GenEditState ges = PackedScene::GEN_EDIT_STATE_INSTANCE;
```
Every scenario where you are instancing the PackedScene should be converted into something that calls [SceneState.get_node_groups](http://docs.godotengine.org/en/latest/classes/class_scenestate.html#class-scenestate-get-node-groups). It should never be actually instancing all of the scenes in the project (that could be colossally slow on larger projects).

xDGameStudios replied:
Hmmm I get you.. but then how can I change and remove nodes from groups. I should only instance the ones that are needed is that what you are saying? I'll get to it now!

xDGameStudios replied:
I can't actually edit node groups using SceneState only. I have to instantiate it, unfortunately

Zylann commented:
I'm worried that deleting all nodes from a deleted group would be a bit too much, could it simply remove them from the group instead? Or at least ask the user if they want that group deleted / and if nodes must also be deleted?

willnationsdev commented:
@Zylann Yeah, I'd recommend avoiding deleting actual nodes. As a group editor, it should strictly be manipulating the group status of nodes within the scenes.

xDGameStudios commented:
> I'm worried that deleting all nodes from a deleted group would be a bit too much, could it simply remove them from the group instead? Or at least ask the user if they want that group deleted / and if nodes must also be deleted? If the action is not undoable it should also be indicated.

You miss understood me, I'm not deleting the nodes, "this will also remove every node in every scene from that group" not remove the node from the project.. but removing the nodes from the deleted group!

CowThing commented:
Does this remove the ability to add non-global groups to nodes in the editor? There might be times when I want to add a group to a node but it doesn't need to be something that's available across the whole project.

Also the autocomplete should probably affect Node.is_in_group() as well.

willnationsdev commented:
@CowThing It probably can't distinguish between project groups and individual groups cleanly. The interface for adding/removing groups is changed to check boxes rather than records you can add, so you'd have to modify the global records in order to make it an option I think.

Perhaps the group tab in the node dock could have an option to add a record to the EditorGroupSettings and reload stuff?

willnationsdev commented:
Also, I'm not sure how useful the "drag and drop" features of the Editor*Settings classes will be for this since the node dock already more-or-less lets you just click on the things you want. You could probably omit those methods altogether.

xDGameStudios commented:
> Also the autocomplete should probably affect Node.is_in_group() as well.

Well though going to add that!

xDGameStudios commented:
The current implementation is open to improvements and I was looking at removing the groups tab in the project settings and combine it with the Group Dock! This however raises some issues:

CONS:

1) I would need to show Group data inside the Group Dock even if no node is selected. The Group Dock lays within the Node Dock, it would be strange to have stuff there even if no node is selected

2) Adding/removing global groups is a BIG thing, imagine you could delete by mistake a group (all scenes with nodes contained in that group would be removed from that group - opened and closed ones)

PROS:

1) Easier access to group management (it would be just there)

2) No need to go back and forth between the project settings window and the node editor.

willnationsdev commented:
> The Group Dock lays within the Node Dock, it would be strange to have stuff there even if no node is selected

I share this concern. Putting editor-wide data in a place that is designed to conditionally show information specific to a single Node seems to me like a no-no (though).

> Adding/removing global groups is a BIG thing, imagine you could delete by mistake a group

This could be solved easily enough by just having a window prompt ask for confirmation so people don't do it accidentally. Might be a good idea even for the EditorGroupSettings version.

> Easier access to group management (it would be just there)
> No need to go back and forth between the project settings window and the node editor.

Big ++ here, and upon first impression, I really love this idea (no redundancy), but the cost from con 1 seems to be too high.

---

I think I'd recommend keeping it the way it is, but adding in redundancies to the node dock's group tab so that people can insert new EditorGroupSettings records directly from the group tab (that way we can gain pro 2 without moving all the data there.

It sucks since it adds more content to the ProjectSettings (which is starting to get pretty full of tabs at this point), but it just makes the most sense to keep the data there.

WDYT @CowThing ?

CowThing commented:
I do think it makes more sense for the group editor to be in project settings, since it's global to the whole project. I don't mind if I have to open it every time I want to add a new global group, because it will only need to be opened once any time I add new groups. It would be weird if the Node Dock had things in it even when no node is selected. I also think the popup to confirm deleting a group is a good idea.

The only thing I don't really like is removing the ability for adding "local" groups in the editor that don't appear globally. But the more I think about it, the more I can't really think of any reason I'd need to make "local" groups like that. Any groups I do use are probably going to be used throughout the project, or are going to be checked via Node.is_in_group() throughout the project, so I would want them to be global. And non-global groups can still be added through scripts which is probably where they're the most use anyways.

willnationsdev commented:
Then I guess we can test it out with just leaving it as it is (plus the dialogue warning for deleting a group), and then if people find that it's too unwieldy to need to pop open the group manager to add stuff, they can create a separate Issue for that.

Zylann commented:
> the more I think about it, the more I can't really think of any reason I'd need to make "local" groups like that

Here is one: plugins.

willnationsdev commented:
@Zylann 

> Here is one: plugins.

What do you mean? If you mean having a plugin make use of groups that aren't established in the main scene, then we could create an API that lets EditorPlugins add/remove records (so that it can toggle their existence when add/removed from the tree). Would that fix the problem?

CowThing commented:
If a plugin is using groups internally for whatever the plugin is trying to accomplish then it's probably best if the plugin adds and removes nodes from groups in the scripts, since there's no reason for the user to ever need to add or remove the groups manually. And if the plugin does need the user to add nodes to groups throughout their project it's probably best if those groups are global so that they are more easily added.

willnationsdev commented:
@CowThing 

> If a plugin is using groups internally for whatever the plugin is trying to accomplish then it's probably best if the plugin adds and removes nodes from groups in the scripts since there's no reason for the user to ever need to add or remove the groups manually.

And that can already be done with the current implementation, so there's no more work to be done (yay).

> And if the plugin does need the user to add nodes to groups throughout their project it's probably best if those groups are global so that they are more easily added.

So then yeah, the EditorPlugin can add/remove the global groups that it expects users to use. That should be an easy addition @xDGameStudios.

willnationsdev commented:
@xDGameStudios @Zylann @CowThing 

Do you guys think we even need this to let people reorder the global groups? I'd probably prefer it if the cached groups could be ordered alphabetically and then generate the Tree from there. Otherwise, I'd probably just be wasting time clicking and clicking trying to move things to be alphabetical and easy to find. You guys opposed to just taking that feature out of this tab completely?

Zylann commented:
@willnationsdev the order of groups doesn't really matter, so better have them sorted in a useful way, that would make sense

xDGameStudios commented:
Added:

- EditorPlugin can add/remove groups
- Groups are now save in the project settings as an array of groups.. to allow storing additional information for future features related to groups.

but right now I have a conflict with the master branch... dont know how to fix it... I'm new to git! but it was a PR that was accepted correcting a grammatical error in code I've deleted because it became obsolete.

Zylann commented:
@xDGameStudios I thought about this usage as well, but I think it's a bit out of scope for now.

willnationsdev commented:
@xDGameStudios @Zylann I would say that having the interface functionality would be out of the scope of this PR specifically (wouldn't want problems with the interface thing to result in the rest of this not getting merged, cause it's good stuff). But, that still leaves the question of whether that feature should be implemented. We should probably open up a separate Issue to discuss that implementation possibility and reference this PR as a basis for it.

DevMagicLord commented:
Tags or groups should be accessible and as easy to use like this 

https://docs.unity3d.com/uploads/Main/TagDropdown.png

willnationsdev commented:
@DevMagicLord that's more or less what the PR does.

xDGameStudios commented:
This used to compile and now something is wrong... though it compiles in my end it gives an error on Travis! @Remy some people said to tag you, and ask if you could take a look at the compiling job. Thank you so much 😊

akien-mga commented:
Restarted CI, sorry for the delay.

Lucrecious commented:
I've got a suggestion!

Instead of putting the groups in an array, it might be nicer to put them in a set, and allow access to them similar to the pretty `get_node('NodeName')` syntax, which is `$NodeName`.

So instead of `var nodes = get_nodes_in_group('GroupName')`, something like `var nodes = &GroupName` to avoid using strings. It would help a lot more with editor warnings since the editor would be able to parse `&GroupName` as `get_nodes_in_group('GroupName')` and check if the group actually exists (similar to `get_node` and the `$` syntax).

Even putting the group names in an easily accessible dictionary, so I can do something like `var nodes = get_nodes_in_group(Group.GroupName)` or `var nodes = get_nodes_in_group(&GroupName)` where `&` would return the actual string value or some hash.

I'd really just like to avoid strings as much as possible. Strings are extremely difficult to refactor later on, and I feel like I'm stuck using a bad group name late into a project because it becomes really difficult to look through all the strings in code to distinguish between a regular string and the actual group name.

Maybe a way to create typed strings is in order. GDScript can mimic the typed strings in other languages but instead only user defined. i.e. instead of just `'GroupName'` maybe `G'GroupName'`. That way at a glance and by grep search I can find all instances where `GroupName` is used as a `GroupName` and not a variable or other type of string.

willnationsdev commented:
@Lucrecious 

> So instead of var nodes = get_nodes_in_group('GroupName'), something like var nodes = &GroupName to avoid using strings. It would help a lot more with editor warnings since the editor would be able to parse &GroupName as get_nodes_in_group('GroupName') and check if the group actually exists (similar to get_node and the $ syntax).

Has already been asked (#25139) and the reception was not good (with good reason).

> Even putting the group names in an easily accessible dictionary, so I can do something like `var nodes = get_nodes_in_group(Group.GroupName)`

This is something I could see happening more readily. If there were some kind of `Groups` singleton that used procedurally generated properties to represent the various groups as strings, then you would have identifier-based access points for the project-wide groups. The same singleton could then also be used to, for example, fetch a list of all groups.

Perhaps we could submit a proposal for that idea *after* this PR gets merged.

CowThing commented:
Despite what I said earlier about not seeing a use for groups that are local to a single scene. I did eventually run into a situation where I used a local group. I had several nodes throughout the scene, all of them down different branches, and I needed to manipulate all of them. But there was no easy way to loop through them since they were all split apart. The only way I could think of was to make a group and put all those nodes in it, then loop through that group. The nodes didn't have scripts on them, so if I wanted to do this through code I would have had to add a script to all of them just to add them to a group, and they weren't all the same type of node.

So perhaps local groups should still exist. It might be better if the group editor looked more like it currently does in Godot, the ability to make new groups and add groups to any nodes in the scene. But also in the list of groups have a section that shows the globally available groups, the groups in this section can't be edited, just added or removed from nodes in the scene.

xDGameStudios commented:
> Despite what I said earlier about not seeing a use for groups that are local to a single scene. I did eventually run into a situation where I used a local group. I had several nodes throughout the scene, all of them down different branches, and I needed to manipulate all of them. But there was no easy way to loop through them since they were all split apart. The only way I could think of was to make a group and put all those nodes in it, then loop through that group. The nodes didn't have scripts on them, so if I wanted to do this through code I would have had to add a script to all of them just to add them to a group, and they weren't all the same type of node.
> 
> So perhaps local groups should still exist. It might be better if the group editor looked more like it currently does in Godot, the ability to make new groups and add groups to any nodes in the scene. But also in the list of groups have a section that shows the globally available groups, the groups in this section can't be edited, just added or removed from nodes in the scene.

Well you could still do that with my project-wide groups.. just create the group in the "Groups" tab... then add whatever nodes you want to it... the diference is that the group is available in every scene you open... so project-wide groups can be used as local... the big advantage of project-wide groups is that when using building a scene at runtime that is composed of multiple individual scenes with nodes that have different the same "group name" you don't have to retype the group name every time for each new scene. And the the parent scene will also automatically be aware of the existing groups in your project.

for example:
```
Player.tscn (base node has group "damage")
Enemy.tscn (base node has group "damage" ... you have to re-create a local group for this)
Tree.tscn (base node has group "damage" ... you have to re-create a local group for this)

Wolrd.tscn (dynamically adds previous scenes to it and calls function on all nodes from "damage" group... you would have to remember the group name is "damage".. and not "Damage" nor "DMG").
```

Another idea of mine would be to make it so groups be defined with implementation requirements... for example if you added Enemy to the "damage" group it could require the node script to have a `take_damage()` and `inflict_damage()` methods (if not an warning would be shown in the scene tree view). This would be something similar to interfaces in other languages.

willnationsdev commented:
@xDGameStudios 

> Well you could still do that with my project-wide groups.. just create the group in the "Groups" tab... then add whatever nodes you want to it. The diference is that the group is available in every scene you open. So project-wide groups can be used as local.

The problem with this is that, assuming @CowThing only wants the nodes inside of that scene to be manipulated, then he needs someway of especially marking that a group *is* local to the scene. Otherwise, if he instantiates that scene multiple times, then the iteration will go over every single node in that group in OTHER scenes in addition to the ones in the local scene. I'm guessing he doesn't want that though.

He basically wants a Godot editor UI-based means of assigning a select subset of nodes in the scene to an Array that can be iterated over. Groups are a good solution for that, but you'd have to somehow define that they are local, not global, and then enable the API to support accessing only scene-local nodes.

The simplest way I can think of doing this would be to append the local scene root's `filepath` property to the group name and use that to limit the group access. Like, create a subsection for local groups, and then have all groups added there use the `<group_name>:<scene_path>` as the group name in the SceneTree behind the scenes. Then you could add a utility method in the Node class, `get_local_nodes_in_group()`, which automatically finds and appends that scene file's path to the provided group to successfully find the local group elements in the SceneTree.

Otherwise...yeah, you'll just have to rely on global groups (the current proposal) or add the nodes to a similarly scene-local group via script.

akien-mga commented:
Sorry for not being to review this yet, the PR backlog is not giving me much bandwidth.

The idea seems interesting, but since we're now very close to the 3.2 release, I'll move this to the next milestone. If it's merged for 4.0 and easy to backport, we might consider adding it to a 3.2.x release.

aaronfranke commented:
@xDGameStudios Is this still desired? If so, it needs to be rebased on the latest master branch.

If not, abandoned pull requests will be closed in the future as announced [here](https://godotengine.org/article/headsup-vulkan-merged-master-branch-unstable).

willnationsdev commented:
@aaronfranke I believe it is still desired as there are quite a number of other issues that have, at this point, referenced it as a solution to other problems. I can copy his branch to my fork and do the rebasing to get it up to speed.

akien-mga commented:
Why isn't the Groups Manager from #16502 sufficient? CC @Paulb23

Zylann commented:
@akien-mga because (from the description), it requires you to already have nodes with a group in your current scene. Meaning if you make a new scene you have to start again adding them manually before they get listed in the dialog.

aaronfranke commented:
Unfortunately, this PR has not received any new commits for over a year and is abandoned, closing.

This feature seems to still be desired, anyone is welcome to salvage it and rebase it.

willnationsdev commented:
Just as an FYI for folks, I've rebased this PR onto a recent master. A lot of the new functionality isn't being called yet (cause the rebase process removed the signal connections and I'd need to figure out what even should be salvaged/used with what's currently in there), but the code is up there (probably doesn't even compile yet). Just noting so that people know most of the code is preserved in another repository.
